### PR TITLE
Makes .py output black compliant

### LINKED ISSUE #482: Auto-generated write_to file triggering errors in flake8, pydocstyle, and pylint
I use the `write_to` option in my `pyproject.toml` config file to write to a Python `__version__.py` file.  It creates a file that looks like this:

```python
# coding: utf-8
# file generated by setuptools_scm
# don't change, don't track in version control
version = '0.1.dev41+ge5eee1b.d20201024'
```

---

**`flake8`**

I run [`flake8`](https://github.com/PyCQA/flake8) (PEP8 linter) on my package with the [`flake8-black`](https://github.com/peterjc/flake8-black) plugin for [`black`](https://github.com/psf/black).

Unfortunately, `flake8` fails with this message:

```none
__version__.py:4:11: BLK100 Black would make changes.
```

I have since fixed things on my side using the [`exclude`](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-exclude) configuration option for `flake8`.

---

**`pydocstyle`**

The same is true for [`pydocstyle`](https://github.com/PyCQA/pydocstyle/):

```none
__version__.py:1 at module level:
        D100: Missing docstring in public module
```

I think we can ignore `pydocstyle`, it seems superfluous to have auto-generated files be given module-level docstrings.

---

**`pylint`**

And also for [`pylint`](https://github.com/PyCQA/pylint):

```none
__version__.py:4:0: C0103: Constant name "version" doesn't conform to UPPER_CASE naming style (invalid-name)
```

I can work around the `pylint` "invalid-name" error using the [`ignore` configuration option](https://github.com/PyCQA/pylint/blob/master/examples/pylintrc#L13).

---

**The Request**

Can the auto-generated files be made compatible with PEP8 (and by the transitive property, `black`)?

### PR DESCRIPTION
**Pull request by ns-rse**
Closes #482

* Replaces single with double quotes in output.
* Updates tests in light of this.

### CODE REVIEW

File: src/setuptools_scm/__init__.py [hunk 1]
```diff
@@ -46,22 +46,27 @@ def dump_version(
...
             )
         )
     version_tuple = _version_as_tuple(version)
 
-    with open(target, "w") as fp:
-        fp.write(template.format(version=version, version_tuple=version_tuple))
+    with target.open("w") as fp:
+        fp.write(
+            template.format(version=version, version_tuple=version_tuple).replace(
+                "'", '"'
```
Comment (RonnyPfannschmidt) [inline src/setuptools_scm/__init__.py]: This unfortunately is unacceptable as it may break custom templates

**Review by RonnyPfannschmidt — CHANGES_REQUESTED** [review]

**Comment by RonnyPfannschmidt** [comment]
Please avoid repeatedly creating new pull requests for the same set of changes

**Comment by ns-rse** [comment]
> Please avoid repeatedly creating new pull requests for the same set of changes

My apologies, I pushed the change too early with an error then cocked up fixing it.

I've now restricted quote substitution to only files that are `.py` which are those that may be linted by Black. I was unsure whether to restrict further to only do so when the `TEMPLATES` derived `.py` are used and would happily take advice on this.

File: src/setuptools_scm/__init__.py [hunk 2]
```diff
@@ -46,21 +46,25 @@ def dump_version(
...
-                os.path.splitext(target)[1], target
+                ext,
+                target,
             )
         )
     version_tuple = _version_as_tuple(version)
 
-    with open(target, "w") as fp:
+    if ext == ".py":
+        template = template.replace("'", '"')
```
Comment (RonnyPfannschmidt) [inline src/setuptools_scm/__init__.py]: this still breaks 3rd party templates while also not solving the repr issue

**Comment by ns-rse** [comment]
Sorry, forgot to check the tests will resolve that now.

File: src/setuptools_scm/__init__.py (hunk 2 above)
Comment (ns-rse) [reply src/setuptools_scm/__init__.py]: Yes I realised when the tests failed (putting child to bed distracted me). Just worked through and have corrected and checked, tests pass locally. :crossed_fingers:

File: src/setuptools_scm/__init__.py [hunk 3]
```diff
@@ -46,22 +46,31 @@ def dump_version(
...
         )
     version_tuple = _version_as_tuple(version)
 
-    with open(target, "w") as fp:
-        fp.write(template.format(version=version, version_tuple=version_tuple))
+    if ext == ".py":
+        with target.open("w") as fp:
+            fp.write(
+                template.format(version=version, version_tuple=version_tuple).replace(
+                    "'", '"'
```
Comment (RonnyPfannschmidt) [inline src/setuptools_scm/__init__.py:68]: this pattern still can break examples

all it needs is a example with a required string thats getting broken by this

this replacement is only safe if it validates before and after generates the same ast

its far safer to have a  str wrapper that adds a format type thats returns a black safe string for tuples or strs

Comment (ns-rse) [reply src/setuptools_scm/__init__.py:68]: Thanks for the explanation and pointers. "_str wrapper_" is something I'm going to have to go and learn about, probably best closing this for now and I'll make another PR when I've understood the suggestion, got some working code and written tests if required.

**Pull request closed without merging**